File: sol/1905/1905.py
class Solution:
    def countSubIslands(self, grid1: List[List[int]], grid2: List[List[int]]) -> int:
        m = len(grid1)
        n = len(grid1[0])
        res = 0
        for i in range(m):
            for j in range(n):
                if grid2[i][j] == 0 or grid1[i][j] == 0:
                    continue
                st = [(i, j)]
                allCovered = True
                while len(st) != 0:
                    a, b = st.pop()
                    if grid2[a][b] == 0:
                        continue
                    grid2[a][b] = 0
                    if grid1[a][b] == 0:
                        allCovered = False
                    if a > 0:
                        st.append((a - 1, b))
                    if a < m - 1:
                        st.append((a + 1, b))
                    if b > 0:
                        st.append((a, b - 1))
                    if b < n - 1:
                        st.append((a, b + 1))
                if allCovered:
                    res += 1
        return res

File: sol/1905/1905.cpp
struct Solution {
    int countSubIslands(vector<vector<int>>& grid1, vector<vector<int>>& grid2) {
        int m = grid1.size(), n = grid1[0].size(), res = 0;
        for (int i = 0; i < m; ++i){
            for (int j = 0; j < n; ++j){
                if (grid2[i][j] == 0 || grid1[i][j] == 0){
                    continue;
                }
                vector<pair<int, int>> st;
                st.push_back({i, j});
                bool allCovered = true;
                while (!st.empty()){
                    auto [a, b] = st.back();
                    st.pop_back();
                    if (grid2[a][b] == 0){
                        continue;
                    }
                    grid2[a][b] = 0;
                    if (grid1[a][b] == 0){
                        allCovered = false;
                    }
                    if (a > 0 && grid2[a - 1][b]){
                        st.push_back({a - 1, b});
                    }
                    if (a < m - 1 && grid2[a + 1][b]){
                        st.push_back({a + 1, b});
                    }
                    if (b > 0 && grid2[a][b - 1]){
                        st.push_back({a, b - 1});
                    }
                    if (b < n - 1 && grid2[a][b + 1]){
                        st.push_back({a, b + 1});
                    }
                }
                res += allCovered;
            }
        }
        return res;
    }
};
